samples/bluetooth/unicast_audio_client/src/main.c

```diff
@@ -37,39 +37,135 @@ static K_SEM_DEFINE(sem_stream_qos, 0, 1);
+ * One sample data buffer en generated and repeated. */
+#define AUDIO_SAMPLE_RATE_HZ    16000
+#define AUDIO_FREQUENCY_HZ      400
+#define AUDIO_LENGTH_US         10000 /* amount of sample data - shall match LC3 frame length */
+#define AUDIO_LENGTH_100US      (AUDIO_LENGTH_US/100)
+#define NUM_SAMPLES             ((AUDIO_LENGTH_US * AUDIO_SAMPLE_RATE_HZ) / 1000000)
+#define AUDIO_VOLUME		(INT16_MAX-3000) /* codec causes clipping above INT16_MAX-3000 */
+
+static int16_t audio_buf[NUM_SAMPLES];
+static lc3_encoder_t lc3_encoder;
+static lc3_encoder_mem_48k_t lc3_encoder_mem;
+
+/**
+ * Use the math lib to generate a sine-wave using 16 bit samples into a buffer.
+ *
+ * @param buf Destination buffer
+ * @param length_us Length of the buffer in microseconds
+ * @param frequency_hz frequency in Hz
+ * @param sample_rate_hz sample-rate in Hz.
+ */
+static void fill_audio_buf_sin(int16_t *buf, int length_us, int frequency_hz, int sample_rate_hz)
 {
-	while (len-- != 0) {
-		printk("%02x", *ptr++);
+	const int sine_period_samples = sample_rate_hz / frequency_hz;
+	const int num_samples = (length_us * sample_rate_hz) / 1000000;
+	const float step = 2 * 3.1415 / sine_period_samples;
+
+	for (int i = 0; i < num_samples; i++) {
+		float sample = sin(i * step);
+
+		buf[i] = (int16_t)(AUDIO_VOLUME * sample);
 	}
 }
 
-static void print_codec(const struct bt_codec *codec)
+static void lc3_audio_timer_timeout(struct k_work *work)
 {
-	printk("codec 0x%02x cid 0x%04x vid 0x%04x count %u\n",
-	       codec->id, codec->cid, codec->vid, codec->data_count);
+	/* For the first call-back we push multiple audio frames to the buffer to use the
+	 * controller ISO buffer to handle jitter.
+	 */
+	const uint8_t prime_count = 2;
+	static int64_t start_time;
+	static int32_t pdu_cnt;
+	int32_t pdu_goal_cnt;
+	int64_t uptime, run_time_ms, run_time_100us;
+	int ret;
+	struct net_buf *buf;
+	uint8_t *net_buffer;
+	uint16_t lc3_ret;
+	const uint16_t tx_sdu_len = audio_stream.iso->qos->tx->sdu;
 
-	for (size_t i = 0; i < codec->data_count; i++) {
-		printk("data #%zu: type 0x%02x len %u\n",
-		       i, codec->data[i].data.type,
-		       codec->data[i].data.data_len);
-		print_hex(codec->data[i].data.data,
-			  codec->data[i].data.data_len -
-				sizeof(codec->data[i].data.type));
-		printk("\n");
+	k_work_schedule(&audio_send_work, K_MSEC(10));
+
+	if (start_time == 0) {
+		/* Read start time and produce the number of frames needed to catch up with any
+		 * inaccuracies in the timer. by calculating the number of frames we should
+		 * have sent and compare to how many were actually sent.
+		 */
+		start_time = k_uptime_get();
 	}
 
-	for (size_t i = 0; i < codec->meta_count; i++) {
-		printk("meta #%zu: type 0x%02x len %u\n",
-		       i, codec->meta[i].data.type,
-		       codec->meta[i].data.data_len);
-		print_hex(codec->meta[i].data.data,
-			  codec->meta[i].data.data_len -
-				sizeof(codec->meta[i].data.type));
-		printk("\n");
+	uptime = k_uptime_get();
```

Comment:
Keep in mind that `k_uptime_get()` uses millisecond units. If sending every 7.5ms, you may get values different than expected. There are APIs to get ticks and to get get microseconds update from ticks if you want a higher resolution than milliseconds.

Reply:
ms is just fine - the jitter on the processing is +/- ~10ms anyway. As the default tick configuration is 10ms I doubt going through ticks will improve resolution even though the unit is us.